src/Scheduler/CronDefinition.php

```diff
@@ -0,0 +1,18 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tempest\Console\Scheduler;
+
+use DateInterval;
+
+final class CronDefinition
+{
+    public function __construct(
```

Comment:
Could you explain these four option? In my mind, we only need an interval; the process will always run in the background, and I don't think output should be configurable on this level (maybe on config level it should)

Reply:
Yeah, I kinda was wondering about the approach we'd want to take here.

Generally speaking if we're going with "the framework that gets out of your way" narrative, we'd probably want to drop this class altogether and just use the DateInterval directly in command method's declaration, the problem I see is that commands are very different from each other and the way people use them will vary.

Some people will use them to just grab some DB records and dispatch a bunch of jobs, some people will use them to modify a status of an entity, where they have one command per status i.e InitializePaymentsCommand, FetchPaymentStatusCommand, SendPaymentConfirmationCommand, and they'd want to have them run in order - not really a fan of that solution myself, but I've seen that in several projects.

What I didn't think about is that wouldn't even be a thing in tempest because we cannot specify order with discovery, so that's not really valid 😅

For other things though - like being able to change an output of specific command to another file, or doing some other customizations may come in handy in the future.

So I'd suggest keeping this class, so we can extend the scheduler without breaking the api, seeing how things like output, or whether or not job could run cross-servers (if that should even be a concern in Tempest, this really depends on the bigger picture) could be important to modify on per-command basis.

Idk how you feel about this, and this def has a hacky feeling to it, but maybe it'd be nice to separate the concept of running something in intervals as a first party citizien.

This way, if we have some simple prunner in our service, we could simply do

```php

#[Interval(minute: 1)]
public function prune(): void
{
   DB::table('messages')->where('created_at', '<=', now())->delete();
}

``` 

This has it's ups and downs and creates more magic but it feels like a nice utility

Reply:
> and they'd want to have them run in order

This shouldn't be a concern of the scheduler; there are other patterns more suited to handling those. One that comes to mind are [sagas](https://learn.microsoft.com/en-us/azure/architecture/reference-architectures/saga/saga) — which are definitely out of scope for the projects Tempest is targeting.

> may come in handy in the future.

It could be, but right now we definitely not going down that route, so it's adding quite a lot of boilerplate for something that might happen. 

> Idk how you feel about this, and this def has a hacky feeling to it, but maybe it'd be nice to separate the concept of running something in intervals as a first party citizien.

Why does it have a hacky feeling? I think your last code block is the way to go, although I'd change the name to `Schedule` for clarity

Reply:
Didn't know about Sagas, thanks for a nice read!

Why does it have a hacky feeling?

Well, what I wanted to do is create this syntax:

```php
#[Interval(minute: 1)]
public function prune(): void
{
   DB::table('messages')->where('created_at', '<=', now())->delete();
}
```

instead of 

```php
#[ConsoleCommand(name: 'abc', schedule: new DateTimeInterval('1Y'))
public function prune(): void
{
   DB::table('messages')->where('created_at', '<=', now())->delete();
}
```

which means that we'd have to be able to call a command from shell to execute that specific handler, but that wouldn't necessarily be a command, just something that runs every X time

Reply:
What I gathered from the PR, the syntax now looks like this

```php
#[Schedule(new DateInterval('P1D'))]
```

I agree that this isn't ideal. Let's brainstorm some options


---

```php
#[Schedule(new DateInterval('P1D'))]
```

This is the most flexible, but also the most confusing syntax

---

```php
#[Schedule(days: 3, hours: 1, minutes: 2)] // At least 3 days, 1 hour and 2 minutes have to be passed
```

A lot more comprehensible, we could even have an option here to run a command on a fixed time:

```php
#[Schedule(at: '10:20')]
```

---

```php
#[Schedule(cron: '3-59/15 6-12 */15 1 2-5')]
```

There are packages that parse cron expressions: https://github.com/dragonmantank/cron-expression

However, while super flexible, I don't think it's the most straight forward.

---

```php
#[Schedule(Every::minute(2))]
```

Not sure about it, just trying to come up with some other approaches.

Reply:
Okay, I just want to clear a couple of things, so we have a shared understanding of the intended functionality.

> which means that we'd have to be able to call a command from shell to execute that specific handler, but that wouldn't necessarily be a command, just something that runs every X time

What I proposed is being able to add that attribute over *any* function in Tempest project, not just commands.

If we want this to work for any function, how would you approach handling that?

One approach would be to serialize the function and use the hash as a name to call from the scheduler, the only downside is that modifications would make it

If we want to make this work **only for commands**, is there a need to add a separate attribute for it, or could this just be another property on `ConsoleCommand`?